class.c

```diff
@@ -37,87 +37,118 @@
 
 RUBY_EXTERN rb_serial_t ruby_vm_global_cvar_state;
 
-void
-rb_class_subclass_add(VALUE super, VALUE klass)
+static void
+push_subclass_entry_to_list(VALUE super, VALUE klass)
 {
     rb_subclass_entry_t *entry, *head;
 
     if (super && super != Qundef) {
-	entry = ALLOC(rb_subclass_entry_t);
+        entry = ZALLOC(rb_subclass_entry_t);
 	entry->klass = klass;
-	entry->next = NULL;
 
 	head = RCLASS_SUBCLASSES(super);
-	if (head) {
-	    entry->next = head;
-	    RCLASS_PARENT_SUBCLASSES(head->klass) = &entry->next;
-	}
+        if (!head) {
+            head = ZALLOC(rb_subclass_entry_t);
+            RCLASS_SUBCLASSES(super) = head;
+        }
+        entry->next = head->next;
+        entry->prev = head;
 
-	RCLASS_SUBCLASSES(super) = entry;
-	RCLASS_PARENT_SUBCLASSES(klass) = &RCLASS_SUBCLASSES(super);
+        if (head->next) {
+            head->next->prev = entry;
+        }
+        head->next = entry;
+
+        if (BUILTIN_TYPE(super) == T_MODULE) {
+            RUBY_ASSERT(BUILTIN_TYPE(klass) == T_ICLASS);
+            RCLASS_MODULE_SUBCLASS_ENTRY(klass) = entry;
+        }
+        else {
+            RUBY_ASSERT(BUILTIN_TYPE(super) == T_CLASS);
+            RCLASS_SUBCLASS_ENTRY(klass) = entry;
+        }
```

Comment:
Why not just keep these in their respective functions (`rb_class_subclass_add` and `rb_module_add_to_subclasses_list`)?

Reply:
I had thought that adding the entry to the front of the list and updating the list pointer were closely related enough that they should be done in the same place, but on reflection I think it makes the code neater to update the list separately and then update the pointer in the original functions. I'll push that up as a change.